sql/core/src/main/scala/org/apache/spark/sql/execution/command/FileWritingCommand.scala

```diff
@@ -0,0 +1,108 @@
+ */
+
+package org.apache.spark.sql.execution.command
+
+import org.apache.spark.SparkContext
+import org.apache.spark.sql.{Row, SparkSession}
+import org.apache.spark.sql.catalyst.{CatalystTypeConverters, InternalRow}
+import org.apache.spark.sql.catalyst.plans.logical
+import org.apache.spark.sql.execution.{SparkPlan, SQLExecution}
+import org.apache.spark.sql.execution.datasources.ExecutedWriteSummary
+import org.apache.spark.sql.execution.metric.{SQLMetric, SQLMetrics}
+import org.apache.spark.util.Utils
+
+/**
+ * A logical command specialized for writing data out. `FileWritingCommand`s are
+ * wrapped in `FileWritingCommandExec` during execution.
+ */
+trait FileWritingCommand extends logical.Command {
+
+  // The caller of `FileWritingCommand` can replace the metrics location by providing this external
+  // metrics structure.
+  private var _externalMetrics: Option[Map[String, SQLMetric]] = None
+  private[sql] def withExternalMetrics(map: Map[String, SQLMetric]): this.type = {
+    _externalMetrics = Option(map)
+    this
+  }
+
+  /**
+   * Those metrics will be updated once the command finishes writing data out. Those metrics will
+   * be taken by `FileWritingCommandExec` as its metrics when showing in UI.
+   */
+  def metrics(sparkContext: SparkContext): Map[String, SQLMetric] = _externalMetrics.getOrElse {
+    Map(
+      // General metrics.
+      "avgTime" -> SQLMetrics.createMetric(sparkContext, "average writing time (ms)"),
+      "numFiles" -> SQLMetrics.createMetric(sparkContext, "number of written files"),
+      "numOutputBytes" -> SQLMetrics.createMetric(sparkContext, "bytes of written output"),
+      "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"),
+      "numParts" -> SQLMetrics.createMetric(sparkContext, "number of dynamic part")
+    )
+  }
+
+  /**
+   * Callback function that update metrics collected from the writing operation.
+   */
+  private[sql] def postDriverMetrics(sparkContext: SparkContext, metrics: Map[String, SQLMetric])
+      (writeSummaries: Seq[ExecutedWriteSummary]): Unit = {
+    var numPartitions = 0
+    var numFiles = 0
+    var totalNumBytes: Long = 0L
+    var totalNumOutput: Long = 0L
+
+    writeSummaries.foreach { summary =>
+      numPartitions += summary.updatedPartitions.size
+      numFiles += summary.numOutputFile
+      totalNumBytes += summary.numOutputBytes
+      totalNumOutput += summary.numOutputRows
+    }
+
+    // The time for writing individual file can be zero if it's less than 1 ms. Zero values can
+    // lower actual time of writing when calculating average, so excluding them.
+    val writingTime =
+      Utils.average(writeSummaries.flatMap(_.writingTimePerFile.filter(_ > 0))).toLong
+
+    val metricsNames = metrics.keys.toSeq.sorted
+    val metricsValues = Seq(writingTime, numFiles, totalNumBytes, totalNumOutput, numPartitions)
+    metricsNames.zip(metricsValues).foreach(x => metrics(x._1).add(x._2))
```

Comment:
I meant user as in caller of this function. This function only works when the input satisfies some requirements and silently fails otherwise. I get it that it's a private method, but it's very error prone.

Imagine I want to extend this by adding another metric called `avgNumFilesPerPart`. If I add it to the end of the `metricsValues` Seq, then all metrics will get messed up (because of the ordering).

At the very least, leave a comment saying that the `metricValues` need to be sorted alphabetically (and rename `writingTime` to `avgWritingTime`). But you should rather consider using a `WriteMetrics` class instead of a Map.